crates/typst-pdf/src/image.rs

```diff
@@ -121,83 +124,184 @@ pub fn deferred_image(
+    }
+}
+
+/// Encode an arbitrary raster image with a suitable filter.
+#[typst_macros::time(name = "encode raster image flate")]
+fn encode_raster_flate(
+    image: &DynamicImage,
+    icc_profile: Option<&[u8]>,
+    interpolate: bool,
+) -> EncodedImage {
+    let color_space = to_color_space(image.color());
+
+    // Encode image data in big-endian. The alpha channel is excluded.
+    // TODO: Encode flate streams with PNG-predictor?
+    let (bits_per_component, data) = match image {
+        DynamicImage::ImageLuma8(buf) => (8, deflate(buf.as_raw())),
+        DynamicImage::ImageLumaA8(_) => (8, deflate(image.to_luma8().as_raw())),
+        DynamicImage::ImageLuma16(buf) => {
+            let encoded: Vec<u8> =
+                buf.as_raw().iter().flat_map(|&c| c.to_be_bytes()).collect();
+            (16, deflate(&encoded))
+        }
+        DynamicImage::ImageLumaA16(buf) => {
+            let encoded: Vec<u8> =
+                buf.pixels().flat_map(|&LumaA([l, _])| l.to_be_bytes()).collect();
+            (16, deflate(&encoded))
+        }
+        DynamicImage::ImageRgb8(buf) => (8, deflate(buf.as_raw())),
+        DynamicImage::ImageRgba8(_) => (8, deflate(image.to_rgb8().as_raw())),
+        DynamicImage::ImageRgb16(buf) => {
+            let encoded: Vec<u8> =
+                buf.as_raw().iter().flat_map(|&c| c.to_be_bytes()).collect();
+            (16, deflate(&encoded))
+        }
+        DynamicImage::ImageRgba16(buf) => {
+            let encoded: Vec<u8> = buf
+                .pixels()
+                .flat_map(|px| px.to_rgb().0)
+                .flat_map(|c| c.to_be_bytes())
+                .collect();
+            (16, deflate(&encoded))
+        }
+        DynamicImage::ImageRgb32F(buf) => {
+            let encoded: Vec<u8> =
+                buf.as_raw().iter().flat_map(|&c| c.to_be_bytes()).collect();
+            (32, deflate(&encoded))
+        }
+        DynamicImage::ImageRgba32F(buf) => {
+            let encoded: Vec<u8> = buf
+                .pixels()
+                .flat_map(|px| px.to_rgb().0)
+                .flat_map(|c| c.to_be_bytes())
+                .collect();
+            (32, deflate(&encoded))
+        }
+        // Anything else
+        _ => (8, deflate(image.to_rgb8().as_raw())),
+    };
+
+    let compressed_icc = icc_profile.map(deflate);
+    let alpha = encode_alpha(image);
+
+    EncodedImage::Raster {
+        data,
+        filter: Filter::FlateDecode,
+        color_space,
+        bits_per_component,
+        width: image.width(),
+        height: image.height(),
+        compressed_icc,
+        alpha,
+        interpolate,
+    }
+}
+
+/// Matches an [`image::ColorType`] to [`ColorSpace`].
+fn to_color_space(color: image::ColorType) -> ColorSpace {
+    use image::ColorType::*;
+    match color {
+        L8 | La8 | L16 | La16 => ColorSpace::D65Gray,
+        Rgb8 | Rgba8 | Rgb16 | Rgba16 | Rgb32F | Rgba32F => ColorSpace::Srgb,
+        _ => unimplemented!(),
```

Comment:
Image implicitly (and naively) converts CMYK to RGB and has no variant for it. When the color channel count determined here is written into the PDF, CMYK JPEG streams will have a mismatch with their image stream dictionary.